client/k8s/informer/K8sInformerFactory.go

```diff
@@ -0,0 +1,138 @@
+package informer
+
+import (
+	"github.com/devtron-labs/devtron/pkg/cluster/repository"
+	"go.uber.org/zap"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	kubeinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"sync"
+	"time"
+)
+
+func NewGlobalMapClusterNamespace() map[string]map[string]string {
+	globalMapClusterNamespace := make(map[string]map[string]string)
+	return globalMapClusterNamespace
+}
+
+type K8sInformerFactoryImpl struct {
+	logger                    *zap.SugaredLogger
+	clusterService            repository.ClusterRepository
+	globalMapClusterNamespace map[string]map[string]string
+}
+type K8sInformerFactory interface {
+	GetLatestNamespaceListGroupByCLuster() map[string]map[string]string
+	BuildInformerForSingleCluster(clusterId int)
+}
+
+func NewK8sInformerFactoryImpl(logger *zap.SugaredLogger, clusterService repository.ClusterRepository, globalMapClusterNamespace map[string]map[string]string) *K8sInformerFactoryImpl {
+	informerFactory := &K8sInformerFactoryImpl{
+		logger:                    logger,
+		clusterService:            clusterService,
+		globalMapClusterNamespace: globalMapClusterNamespace,
+	}
+	informerFactory.buildInformer()
+	return informerFactory
+}
+
+func (impl *K8sInformerFactoryImpl) GetLatestNamespaceListGroupByCLuster() map[string]map[string]string {
+	copiedClusterNamespaces := make(map[string]map[string]string)
+	for key, value := range impl.globalMapClusterNamespace {
+		for _, namespace := range value {
+			if _, ok := copiedClusterNamespaces[key]; !ok {
+				allNamespaces := make(map[string]string)
+				allNamespaces[namespace] = namespace
+				copiedClusterNamespaces[key] = allNamespaces
+			} else {
+				allNamespaces := copiedClusterNamespaces[key]
+				allNamespaces[namespace] = namespace
+				copiedClusterNamespaces[key] = allNamespaces
+			}
+		}
+	}
+	return copiedClusterNamespaces
+}
+
+func (impl *K8sInformerFactoryImpl) BuildInformerForSingleCluster(clusterId int) {
+	model, err := impl.clusterService.FindById(clusterId)
+	if err != nil {
+		impl.logger.Errorw("error in fetching cluster", "err", err)
+		return
+	}
+	var mutex sync.Mutex
```

Comment:
no initialised before use. might throw null pointer

Reply:
use single mutex for all read write in a map

Reply:
no initialization is required here for mutex, it will works. using single mutex for all cluster config..